### PR TITLE
Respect Grafana query interval in Spans Stats requests

### PR DESCRIPTION
**Pull request by sossup**
# Summary

Spans Stats queries (`/events-stats/?dataset=spans`) were not forwarding the Grafana query interval to Sentry. As a result, Sentry always fell back to its default rollup, causing the returned time series to use bucket sizes that did not match the panel interval (for example, a panel configured with a 5-minute interval could receive hourly buckets).

# Changes

- Forward the Grafana interval (`intervalMs` → `backendQuery.Interval`) to Sentry as the `interval` query parameter using the existing `FormatSentryInterval` helper.
- Snap the requested interval to the nearest value accepted by the Sentry Spans Stats API, since the API only supports a fixed set of intervals:
  - `15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400` seconds.
  - Examples:
    - `32s` → `30s`
    - `119s` → `120s` (`2m`)
- Omit the `interval` parameter (allowing Sentry to use its default rollup) when:
  - the interval is below `15s`;
  - the interval is above `24h`;
  - the interval is larger than the query time range, since Sentry rejects these requests with `Interval cannot be larger than the date range`.

# Notes / Questions

I'm not a maintainer of this project, so I'd appreciate some guidance on the expected contribution workflow.

- Is this implementation the preferred approach, or would you like it handled differently?
- Since I only need this change for my own Grafana instance, would you recommend keeping it as a local patch instead of contributing it upstream, or is opening a PR like this the preferred approach?

If you'd like this implemented differently, I'm happy to update the PR based on your feedback.

### CODE REVIEW

File: pkg/sentry/spans_stats.go [hunk 1]
```diff
@@ -9,6 +9,29 @@ import (
 
 type SentrySpansStats = map[string]interface{}
 
+var allowedSpansStatsIntervals = []int64{15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400}
+
+func (gei *GetSpansStatsInput) snapInterval() bool {
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: It might be worth adding a table-driven test for this, since the behaviour is all boundary conditions: 14s and 15s, the 32s → 30s example from the description, an equidistant value like 45s (which snaps down to 30s because of the strict `<`), 25h, and an interval larger than the range.

File: pkg/sentry/spans_stats.go [hunk 2]
```diff
@@ -9,6 +9,29 @@ import (
...
+		return false
+	}
+	nearest := min
+	for _, seconds := range allowedSpansStatsIntervals {
+		valid := time.Duration(seconds) * time.Second
+		if (gei.Interval - valid).Abs() < (gei.Interval - nearest).Abs() {
+			nearest = valid
+		}
+	}
+	if nearest > gei.To.Sub(gei.From) {
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: When the nearest value rounds up past the time range, the interval is dropped entirely even if a smaller allowed value would fit (for example a 50s interval over a 55s range snaps to 60s and is omitted, though 30s is valid). Consider snapping down to the largest allowed value that still fits the range.

File: pkg/sentry/spans_stats.go [hunk 3]
```diff
@@ -9,6 +9,29 @@ import (
 
 type SentrySpansStats = map[string]interface{}
 
+var allowedSpansStatsIntervals = []int64{15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400}
+
+func (gei *GetSpansStatsInput) snapInterval() bool {
+	min := time.Duration(allowedSpansStatsIntervals[0]) * time.Second
+	max := time.Duration(allowedSpansStatsIntervals[len(allowedSpansStatsIntervals)-1]) * time.Second
+
+	if gei.Interval < min || gei.Interval > max {
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: Sub-15s intervals fall back to Sentry's default rollup, which can be much coarser than what the panel asked for (Explore's default 1h range yields a 5s interval, so the feature never engages there). Would clamping up to 15s be closer to the intent than omitting?

File: pkg/sentry/spans_stats.go [hunk 4]
```diff
@@ -9,6 +9,29 @@ import (
 
 type SentrySpansStats = map[string]interface{}
 
+var allowedSpansStatsIntervals = []int64{15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400}
+
+func (gei *GetSpansStatsInput) snapInterval() bool {
+	min := time.Duration(allowedSpansStatsIntervals[0]) * time.Second
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: nit: `min` and `max` shadow the Go 1.21+ builtins here. Renaming to something like `minAllowed`/`maxAllowed` would avoid the confusion.

File: pkg/sentry/spans_stats.go [hunk 5]
```diff
@@ -9,6 +9,29 @@ import (
 
 type SentrySpansStats = map[string]interface{}
 
+var allowedSpansStatsIntervals = []int64{15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400}
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: nit: the first/last-element min/max lookup silently relies on this slice staying sorted. A short comment noting that invariant would protect against a careless future addition.

**Review by adamyeats — COMMENTED** [review]
@sossup Thanks for the contribution! I confirmed the logic behaves as described end to end. Please ensure your commits are GPG signed before we can merge this. A few comments below:

File: pkg/sentry/spans_stats.go [hunk 6]
```diff
@@ -9,6 +9,29 @@ import (
...
+	for _, seconds := range allowedSpansStatsIntervals {
+		valid := time.Duration(seconds) * time.Second
+		if (gei.Interval - valid).Abs() < (gei.Interval - nearest).Abs() {
+			nearest = valid
+		}
+	}
+	if nearest > gei.To.Sub(gei.From) {
+		return false
+	}
+	gei.Interval = nearest
```
Comment (adamyeats) [inline pkg/sentry/spans_stats.go]: nit: mutating the receiver and returning a bool makes this a little side-effecty for a query builder, though the `Limit` clamp above sets precedent. Consider returning `(time.Duration, bool)` and leaving the input untouched.

File: pkg/sentry/spans_stats.go (hunk 1 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: Added the table-driven tests as suggested — they cover the 14s/15s boundary, the 32s → 30s case, the 45s equidistant tie (snapping down due to the strict <), 25h snapping to 24h, an interval larger than the range, plus the snap-down-to-fit and shorter-than-minimum-range cases. Let me know if you see any scenario missing.

File: pkg/sentry/spans_stats.go (hunk 2 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: The logic changed here (I'll expand on the reasoning in another comment): the interval is never omitted anymore — the function now always resolves to a valid value. For the case you described, it snaps down to the largest allowed value that still fits the range (50s over a 55s range now returns 30s instead of being dropped). There's a test pinning exactly that example.

File: pkg/sentry/spans_stats.go (hunk 3 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: Some context on the original design: the Grafana version I run always sends an interval computed from time range / max data points, and my self-hosted Sentry was defaulting to very coarse rollups (15m) when the parameter was omitted — that's what I was trying to work around. You're right that omitting made the feature never engage on short ranges. The new logic addresses this: sub-15s intervals now fall back to a range-derived suggestion, which for short ranges effectively clamps up to 15s, and for long ranges avoids Sentry's coarse default.

File: pkg/sentry/spans_stats.go (hunk 4 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: To be honest this is my first time working with Go — I see the concern and it makes total sense not to shadow language builtins. Renamed to minAllowed/maxAllowed as suggested, thanks!

File: pkg/sentry/spans_stats.go (hunk 5 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: Agreed this could bite a future edit. Instead of a comment (the codebase has very few of them), I opted to enforce the invariant in code: the function sorts the slice before using it. For a 15-element list the cost is negligible. If you'd rather keep just the comment, happy to change it.

File: pkg/sentry/spans_stats.go (hunk 6 above)
Comment (sossup) [reply pkg/sentry/spans_stats.go]: Rethinking this together with the omission discussion, I refactored it into a pure function: snapSpansStatsInterval(interval, timeRange) in a new interval.go, no receiver mutation and no boolean. Grafana itself computes the default interval as time_range / max_data_points, but that can be absent on requests made outside the Grafana UI — so I replicated the same idea as an internal fallback: when the incoming interval is unusable (zero, below Sentry's minimum, or larger than the range), the function derives a suggestion from time_range / spansStatsTargetBuckets and then snaps it to the nearest allowed value.

One thing I'd like your opinion on: I set the target to 500 buckets, but Grafana's actual maxDataPoints varies (it's derived from panel width and can be overridden per query), so a fixed 500 may be wrong for some visualizations. Would you use backendQuery.MaxDataPoints here instead, falling back to a constant only when it's absent?

**Comment by sossup** [comment]
## Summary of changes

`FormatSentryInterval` was defined in `events_stats.go` but is now called from three places, and this PR adds a second interval-related function — so I extracted both into a new `interval.go`, acting as a small interval utility for the `sentry` package.

The omission logic was replaced with something less error-prone: instead of dropping the parameter when the requested interval doesn't fit Sentry's accepted values, the function now always resolves to a valid interval. It mirrors the calculation Grafana itself uses (`time_range / max_data_points`) to build a suggestion from the time range, then snaps it to the nearest value Sentry accepts, snapping down when the result wouldn't fit the range. This way the interval is never omitted and we never fall back to Sentry's very coarse server-side defaults.

**Tests:** 10 table-driven cases covering the boundary scenarios (fallbacks, nearest-value snapping up/down, equidistant ties, values above the maximum, ranges shorter than the minimum), which also serve as a safety net for future changes to the function — plus one test guaranteeing `allowedSpansStatsIntervals` is sorted before use.

**Next steps:** I'll wait for feedback on the comments above before deciding how to handle `max_data_points` (whether to keep the fixed bucket target or use `backendQuery.MaxDataPoints`), and once that's settled I'll mark the PR as ready for review.

> First of all, thanks for the quick turnaround on the review — I really appreciate the work you all put into maintaining this project.
> 
> One question, a bit beyond this PR: I've reached a point where I might end up not running the official version of the plugin. I've been relying heavily on Sentry's Uptime feature, even more since updating to 26.5.2 this week and enabling the option to validate the returned JSON. I'm considering building something that consumes Sentry's `/uptime` API and exposes it as a signal in Grafana, so I can finally have a dashboard with everything I want to watch in real time.
> 
> Is Uptime support on your roadmap? If I finish this and take the time to study the project as a whole to follow its conventions, would you be open to reviewing a new PR for it?

**Comment by adamyeats** [comment]
> Is Uptime support on your roadmap? If I finish this and take the time to study the project as a whole to follow its conventions, would you be open to reviewing a new PR for it?

You might be interested in [this](https://github.com/grafana/sentry-datasource/pull/725)! 😄 However, as it stands, this will go into `v3.0.0`. This will be a major version bump and quite a substantial breaking change, as it will make the long-overdue migration to Sentry's newer API, which is almost certainly going to require changes to existing dashboards making certain types of queries. If you can make the jump, then the Uptime query type will be available for you.

I'm not able to give timelines on this at the moment on when this'll be ready, unfortunately. I will review your changes later today, thanks for actioning those.

File: pkg/sentry/spans_stats.go (hunk 6 above)
Comment (adamyeats) [reply pkg/sentry/spans_stats.go]: Yes, I think `backendQuery.MaxDataPoints` is the better source when it's present, since `range / maxDataPoints` is exactly how Grafana derives the interval itself, so the fallback would respect per-panel resolution rather than assuming 500. It needs a small bit of plumbing (a `MaxDataPoints` field on `GetSpansStatsInput`, wired up in `handlers.go` alongside `Interval`) plus a guard to fall back to the constant when it's zero or absent. Happy for that to land in this PR or as a follow-up, whichever you prefer.

File: pkg/sentry/interval.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+var allowedSpansStatsIntervals = []int64{
+	15, 30, 60, 120, 300, 600, 900, 1800, 3600, 7200, 10800, 14400, 21600, 43200, 86400,
+}
+
+const spansStatsTargetBuckets = 500
+
+func snapSpansStatsInterval(interval time.Duration, timeRange time.Duration) string {
+	slices.Sort(allowedSpansStatsIntervals)
```
Comment (adamyeats) [inline pkg/sentry/interval.go:17]: Sorting the shared package-level slice on every call mutates global state, so two concurrent queries could in principle race in exactly the case the sort is guarding against. Sorting once in an `init()` (or `sync.OnceFunc`) would give the same invariant without the per-call mutation, and the companion test could then assert the sorted var directly.

File: pkg/sentry/interval_test.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+func TestSnapSpansStatsIntervalUnsortedList(t *testing.T) {
+	slices.Reverse(allowedSpansStatsIntervals)
+	snapSpansStatsInterval(32*time.Second, time.Hour)
+	if !slices.IsSorted(allowedSpansStatsIntervals) {
+		t.Error("allowedSpansStatsIntervals should be sorted after the call")
+	}
+}
```
Comment (adamyeats) [inline pkg/sentry/interval_test.go]: nit: the file is missing a trailing newline, so `gofmt -l` flags it.

**Review by adamyeats — DISMISSED** [review]
Thanks for the quick turnaround, this all looks good now and the tests pin the tricky boundaries nicely. Two small comments below. Could you move the PR from draft to ready for review when you get a chance?

**Review by adamyeats — APPROVED** [review]
Thanks @sossup for the great work on this and for the careful follow-through on every round of feedback. To answer your earlier question, contributing this upstream was absolutely the right call, and it will benefit everyone using Spans Stats.
